server: server orchestrator can fail to start tenant

Seen [here](https://teamcity.cockroachdb.com/viewLog.html?buildId=12121007&buildTypeId=Cockroach_BazelExtendedCi) on #110932 that shouldn't cause this flake:
```
Failed
=== RUN   TestRevertTenantToTimestamp/reverts_changes_inside_a_tenant
    stream_ingestion_manager_test.go:85: condition failed to evaluate within 45s: pq: service unavailable for target tenant (target)
    --- FAIL: TestRevertTenantToTimestamp/reverts_changes_inside_a_tenant (46.91s)
```
I'm assuming we want to skip this test under stress.

Jira issue: CRDB-32225

--- a/pkg/ccl/streamingccl/streamingest/stream_ingestion_job_test.go
+++ b/pkg/ccl/streamingccl/streamingest/stream_ingestion_job_test.go
@@ -740,6 +740,7 @@ func waitUntilTenantServerStopped(
 		if err != nil {
 			return err
 		}
+		defer func() { _ = db.Close() }()
 		if err := db.Ping(); err == nil {
 			t.Logf("tenant %q is still accepting connections", tenantName)
 			return errors.Newf("tenant %q still accepting connections")

--- a/pkg/ccl/streamingccl/streamingest/stream_ingestion_job_test.go
+++ b/pkg/ccl/streamingccl/streamingest/stream_ingestion_job_test.go
@@ -132,45 +132,6 @@ func TestTenantStreamingFailback(t *testing.T) {
 		return conn
 	}
 
-	// ALTER VIRTUAL CLUSTER STOP SERVICE does not block until the
-	// service is stopped. But, we need to wait until the
-	// SQLServer is stopped to ensure that nothing is writing to
-	// the relevant keyspace.
-	waitUntilStopped := func(t *testing.T, srv serverutils.ApplicationLayerInterface, tenantName string) {
-		// TODO(ssd): We may want to do something like this,
-		// but this query is driven first from the
-		// system.tenants table and doesn't strictly represent
-		// the in-memory state of the tenant controller.
-		//
-		// client := srv.GetAdminClient(t)
-		// testutils.SucceedsSoon(t, func() error {
-		// 	resp, err := client.ListTenants(ctx, &serverpb.ListTenantsRequest{})
-		// 	if err != nil {
-		// 		return err
-		// 	}
-		// 	for _, tenant := range resp.Tenants {
-		// 		if tenant.TenantName == tenantName {
-		// 			t.Logf("tenant %q is still running", tenantName)
-		// 			return errors.Newf("tenant %q still running")
-		// 		}
-		// 	}
-		// 	t.Logf("tenant %q is not running", tenantName)
-		// 	return nil
-		// })
-		testutils.SucceedsSoon(t, func() error {
-			db, err := srv.SQLConnE(serverutils.DBName(fmt.Sprintf("cluster:%s", tenantName)))
-			if err != nil {
-				return err
-			}
-			if err := db.Ping(); err == nil {
-				t.Logf("tenant %q is still accepting connections", tenantName)
-				return errors.Newf("tenant %q still accepting connections")
-			}
-			t.Logf("tenant %q is not accepting connections", tenantName)
-			return nil
-		})
-	}
-
 	sqlA := sqlutils.MakeSQLRunner(aDB)
 	sqlB := sqlutils.MakeSQLRunner(bDB)
 
@@ -287,7 +248,7 @@ func TestTenantStreamingFailback(t *testing.T) {
 	//   Fingerprint f and g as of ts1
 	//   Fingerprint f and g as of ts2
 	sqlA.Exec(t, "ALTER VIRTUAL CLUSTER f STOP SERVICE")
-	waitUntilStopped(t, serverA.SystemLayer(), "f")
+	waitUntilTenantServerStopped(t, serverA.SystemLayer(), "f")
 	t.Logf("starting replication g->f")
 	sqlA.Exec(t, fmt.Sprintf("SELECT crdb_internal.unsafe_revert_tenant_to_timestamp('f', %s)", ts1))
 	sqlA.Exec(t, fmt.Sprintf("CREATE VIRTUAL CLUSTER f FROM REPLICATION OF g ON $1 WITH RESUME TIMESTAMP = '%s'", ts1), serverBURL.String())
@@ -317,7 +278,7 @@ func TestTenantStreamingFailback(t *testing.T) {
 	sqlA.Exec(t, "ALTER VIRTUAL CLUSTER f START SERVICE SHARED")
 
 	sqlB.Exec(t, "ALTER VIRTUAL CLUSTER g STOP SERVICE")
-	waitUntilStopped(t, serverB.SystemLayer(), "g")
+	waitUntilTenantServerStopped(t, serverB.SystemLayer(), "g")
 	t.Logf("starting replication f->g")
 	sqlB.Exec(t, fmt.Sprintf("SELECT crdb_internal.unsafe_revert_tenant_to_timestamp('g', %s)", ts3))
 	sqlB.Exec(t, fmt.Sprintf("CREATE VIRTUAL CLUSTER g FROM REPLICATION OF f ON $1 WITH RESUME TIMESTAMP = '%s'", ts3), serverAURL.String())
@@ -745,4 +706,45 @@ func TestCutoverCheckpointing(t *testing.T) {
 	// the empty remainingSpans are encoded as
 	// roachpb.Spans{roachpb.Span{Key:/Min, EndKey:/Min}.
 	require.Equal(t, getCutoverRemainingSpans().String(), roachpb.Spans{}.String())
+}
+
+// ALTER VIRTUAL CLUSTER STOP SERVICE does not block until the service
+// is stopped. But, we need to wait until the SQLServer is stopped to
+// ensure that nothing is writing to the relevant keyspace.
+func waitUntilTenantServerStopped(
+	t *testing.T, srv serverutils.ApplicationLayerInterface, tenantName string,
+) {
+	t.Helper()
+	// TODO(ssd): We may want to do something like this,
+	// but this query is driven first from the
+	// system.tenants table and doesn't strictly represent
+	// the in-memory state of the tenant controller.
+	//
+	// client := srv.GetAdminClient(t)
+	// testutils.SucceedsSoon(t, func() error {
+	// 	resp, err := client.ListTenants(ctx, &serverpb.ListTenantsRequest{})
+	// 	if err != nil {
+	// 		return err
+	// 	}
+	// 	for _, tenant := range resp.Tenants {
+	// 		if tenant.TenantName == tenantName {
+	// 			t.Logf("tenant %q is still running", tenantName)
+	// 			return errors.Newf("tenant %q still running")
+	// 		}
+	// 	}
+	// 	t.Logf("tenant %q is not running", tenantName)
+	// 	return nil
+	// })
+	testutils.SucceedsSoon(t, func() error {
+		db, err := srv.SQLConnE(serverutils.DBName(fmt.Sprintf("cluster:%s", tenantName)))
+		if err != nil {
+			return err
+		}
+		if err := db.Ping(); err == nil {
+			t.Logf("tenant %q is still accepting connections", tenantName)
+			return errors.Newf("tenant %q still accepting connections")
+		}
+		t.Logf("tenant %q is not accepting connections", tenantName)
+		return nil
+	})
 }

--- a/pkg/ccl/streamingccl/streamingest/stream_ingestion_manager_test.go
+++ b/pkg/ccl/streamingccl/streamingest/stream_ingestion_manager_test.go
@@ -115,6 +115,8 @@ func TestRevertTenantToTimestamp(t *testing.T) {
 		tenantSQL.CheckQueryResults(t, "SELECT max(k) FROM test.revert1", [][]string{{"1100"}})
 
 		systemSQL.Exec(t, "ALTER VIRTUAL CLUSTER target STOP SERVICE")
+		waitUntilTenantServerStopped(t, srv.SystemLayer(), "target")
+
 		systemSQL.Exec(t, fmt.Sprintf("SELECT crdb_internal.unsafe_revert_tenant_to_timestamp('target', %s)", ts))
 		systemSQL.Exec(t, "ALTER VIRTUAL CLUSTER target START SERVICE SHARED")
 

--- a/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go
+++ b/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go
@@ -46,10 +46,11 @@ type Watcher struct {
 	mu struct {
 		syncutil.RWMutex
 
-		store  map[roachpb.TenantID]*watcherEntry
-		byName map[roachpb.TenantName]roachpb.TenantID
+		store      map[roachpb.TenantID]*watcherEntry
+		byName     map[roachpb.TenantName]roachpb.TenantID
+		allTenants []tenantcapabilities.Entry
 
-		// anyChangeChs is closed on any change to the set of
+		// anyChangeCh is closed on any change to the set of
 		// tenants.
 		anyChangeCh chan struct{}
 	}
@@ -166,18 +167,18 @@ func (w *Watcher) GetCapabilities(
 // GetAllTenants returns all known tenant entries and a channel that
 // is closed if any of the tenants change or any tenants are added or
 // removed.
+//
+// TODO(ssd): Memory ownership could be a bit cleaner here. To avoid
+// needlessly allocating the returned slice every time this is called,
+// we return the same slice to all callers and assume they don't do
+// anything problematic with it. At the moment, there is only one
+// caller of this function so this isn't particularly problematic, but
+// it could be if we add more callers.
 func (w *Watcher) GetAllTenants() ([]tenantcapabilities.Entry, <-chan struct{}) {
 	w.mu.RLock()
 	defer w.mu.RUnlock()
 
-	entries := make([]tenantcapabilities.Entry, 0, len(w.mu.store))
-	for _, v := range w.mu.store {
-		if v.Entry != nil {
-			entries = append(entries, *v.Entry)
-		}
-	}
-
-	return entries, w.mu.anyChangeCh
+	return w.mu.allTenants, w.mu.anyChangeCh
 }
 
 // GetGlobalCapabilityState implements the tenantcapabilities.Reader interface.
@@ -302,13 +303,22 @@ func (w *Watcher) handleUpdate(ctx context.Context, u rangefeedcache.Update) {
 	}
 
 	if len(updates) > 0 {
-		w.closeAnyChangeCh()
+		w.onAnyChange()
 	}
 }
 
-func (w *Watcher) closeAnyChangeCh() {
+func (w *Watcher) onAnyChange() {
 	w.mu.Lock()
 	defer w.mu.Unlock()
+
+	entries := make([]tenantcapabilities.Entry, 0, len(w.mu.store))
+	for _, v := range w.mu.store {
+		if v.Entry != nil {
+			entries = append(entries, *v.Entry)
+		}
+	}
+	w.mu.allTenants = entries
+
 	close(w.mu.anyChangeCh)
 	w.mu.anyChangeCh = make(chan struct{})
 }

--- a/pkg/server/server_controller_orchestration.go
+++ b/pkg/server/server_controller_orchestration.go
@@ -12,10 +12,12 @@ package server
 
 import (
 	"context"
+	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/clusterversion"
 	"github.com/cockroachdb/cockroach/pkg/multitenant/mtinfopb"
+	"github.com/cockroachdb/cockroach/pkg/multitenant/tenantcapabilities"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/isql"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/catconstants"
@@ -25,6 +27,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/log/logpb"
 	"github.com/cockroachdb/cockroach/pkg/util/startup"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/redact"
 )
@@ -41,28 +44,33 @@ func (c *serverController) start(ctx context.Context, ie isql.Executor) error {
 
 	// Run the detection of which servers should be started or stopped.
 	return c.stopper.RunAsyncTask(ctx, "mark-tenant-services", func(ctx context.Context) {
+		// We receieve updates from the tenantcapabilities
+		// watcher, but we also refresh our state at a fixed
+		// interval to account for:
+		//
+		//  - A rapid stop & start in which the start is
+		//    initially ignored because the server is still
+		//    stopping.
+		//
+		//  - Startup failures that we want to retry.
+		const watchInterval = 1 * time.Second
 		ctx, cancel := c.stopper.WithCancelOnQuiesce(ctx)
 		defer cancel()
 
+		timer := timeutil.NewTimer()
+		defer timer.Stop()
+
 		for {
 			allTenants, updateCh := c.watcher.GetAllTenants()
-			tenantsToStart := make([]roachpb.TenantName, 0, len(allTenants))
-			for _, e := range allTenants {
-				if e.Name == "" {
-					continue
-				}
-
-				if e.DataState == mtinfopb.DataStateReady && e.ServiceMode == mtinfopb.ServiceModeShared {
-					tenantsToStart = append(tenantsToStart, e.Name)
-				}
-			}
-
-			if err := c.startMissingServers(ctx, tenantsToStart); err != nil {
+			if err := c.startMissingServers(ctx, allTenants); err != nil {
 				log.Warningf(ctx, "cannot update running tenant services: %v", err)
 			}
 
+			timer.Reset(watchInterval)
 			select {
 			case <-updateCh:
+			case <-timer.C:
+				timer.Read = true
 			case <-c.stopper.ShouldQuiesce():
 				// Expedited server shutdown of outer server.
 				return
@@ -103,11 +111,24 @@ func (c *serverController) startInitialSecondaryTenantServers(
 }
 
 func (c *serverController) startMissingServers(
-	ctx context.Context, tenants []roachpb.TenantName,
+	ctx context.Context, tenants []tenantcapabilities.Entry,
 ) error {
 	c.mu.Lock()
 	defer c.mu.Unlock()
-	for _, name := range tenants {
+	for _, t := range tenants {
+		if t.Name == "" {
+			continue
+		}
+
+		if t.DataState != mtinfopb.DataStateReady {
+			continue
+		}
+
+		if t.ServiceMode != mtinfopb.ServiceModeShared {
+			continue
+		}
+
+		name := t.Name
 		if _, ok := c.mu.servers[name]; !ok {
 			log.Infof(ctx, "tenant %q has changed service mode, should now start", name)
 			// Mark the server for async creation.

--- a/pkg/server/server_controller_test.go
+++ b/pkg/server/server_controller_test.go
@@ -19,9 +19,12 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
@@ -67,6 +70,58 @@ func TestServerController(t *testing.T) {
 	// tenant constructor was called.
 	require.Error(t, err, "tenant connector requires a CCL binary")
 	// TODO(knz): test something about d.
+}
+
+// TestServerControllerStopStart is, when run under stress, a
+// regression test for #112077, a bug in which we would fail to
+// respond to a service start request that occured while a server was
+// shutting down.
+func TestServerControllerStopStart(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+
+	s, db, _ := serverutils.StartServer(t, base.TestServerArgs{
+		DefaultTestTenant: base.TestControlsTenantsExplicitly,
+	})
+	defer s.Stopper().Stop(ctx)
+
+	sqlRunner := sqlutils.MakeSQLRunner(db)
+	// Speed up the tenant capabilities watcher to increase chance of hitting race.
+	sqlRunner.Exec(t, "SET CLUSTER SETTING kv.closed_timestamp.target_duration = '100ms'")
+	sqlRunner.Exec(t, "SET CLUSTER SETTING kv.closed_timestamp.side_transport_interval ='100ms'")
+	sqlRunner.Exec(t, "SET CLUSTER SETTING kv.rangefeed.closed_timestamp_refresh_interval ='100ms'")
+
+	tryConnect := func() error {
+		conn, err := s.SystemLayer().SQLConnE(serverutils.DBName("cluster:hello"))
+		if err != nil {
+			return err
+		}
+		defer func() { _ = conn.Close() }()
+		return conn.Ping()
+	}
+
+	shouldConnectSoon := func() {
+		testutils.SucceedsSoon(t, tryConnect)
+	}
+
+	shouldFailToConnectSoon := func() {
+		testutils.SucceedsSoon(t, func() error {
+			if err := tryConnect(); err == nil {
+				return errors.Newf("still accepting connections")
+			}
+			return nil
+		})
+	}
+
+	sqlRunner.Exec(t, "CREATE TENANT hello")
+	sqlRunner.Exec(t, "ALTER VIRTUAL CLUSTER hello START SERVICE SHARED")
+	shouldConnectSoon()
+	sqlRunner.Exec(t, "ALTER VIRTUAL CLUSTER hello STOP SERVICE")
+	shouldFailToConnectSoon()
+	sqlRunner.Exec(t, "ALTER VIRTUAL CLUSTER hello START SERVICE SHARED")
+	shouldConnectSoon()
 }
 
 func TestSQLErrorUponInvalidTenant(t *testing.T) {

